### PR TITLE
Convert Google::Protobuf.deep_copy to pure Ruby

### PR DESCRIPTION
**Pull request by tenderlove**
In general, I think it will help us to debug issues if we have less C
code and more Ruby code.  This method can be implemented in pure Ruby,
so this commit converts it to pure Ruby.

### CODE REVIEW

File: ruby/ext/google/protobuf_c/message.c [hunk 1]
```diff
@@ -512,7 +512,7 @@ VALUE Message_dup(VALUE _self) {
 }
 
 // Internal only; used by Google::Protobuf.deep_copy.
```
Comment (haberman) [inline ruby/ext/google/protobuf_c/message.c]: If we are exposing this method publicly we should remove this comment.

In many cases I would prefer to keep internal-only methods private. But in this case I don't see any benefit in hiding `Google::Protobuf::Map::deep_copy` while `Google::Protobuf::deep_copy` is public.

Comment (tenderlove) [reply ruby/ext/google/protobuf_c/message.c]: 👍 I'll remove the comments

Comment (tenderlove) [reply ruby/ext/google/protobuf_c/message.c]: I changed them to be `:nodoc:`.  We usually `:nodoc:` "publicly visible but private" methods in Rails, and I think it makes sense in this case too.

**Comment by haberman** [comment]
Hmm, this appears to have broken our Ruby Kokoro tests: https://source.cloud.google.com/results/invocations/7f69f595-c945-4d75-92bb-673e2f486878/targets/protobuf%2Fgithub%2Fmaster%2Fubuntu%2Fruby23%2Fpresubmit/log

Not sure why they didn't run when the PR was open. We should roll this back until we figure out the fix.

**Comment by haberman** [comment]
Oh, I think I was supposed to add the "kokoro:run" label. Sorry about that.

**Pull request merged**